File: src/main/java/nieboczek/lifestolen/config/ClientConfig.java
package nieboczek.lifestolen.config;

import nieboczek.lifestolen.serializer.base.*;

import java.util.Map;

/// Access instance at {@link nieboczek.lifestolen.Lifestolen#cfg}
public final class ClientConfig {
    public static final String ID = "client";

    public boolean renderClientBrandText;
    public float textScale = 1f;
    public Map<String, Boolean> enabledModules;

    public static Serializer<ClientConfig> getSerializer() {
        return ClassSerializer.of(ClientConfig::new)
                .field("renderClientBrandText", BooleanSerializer.of(), c -> c.renderClientBrandText, (c, v) -> c.renderClientBrandText = v)
                .field("textScale", FloatSerializer.of(), c -> c.textScale, (c, v) -> c.textScale = v)
                .field("enabledModules", MapSerializer.of(IdentifierSerializer.of(), BooleanSerializer.of()), c -> c.enabledModules, (c, v) -> c.enabledModules = v);
    }
}

File: src/main/kotlin/nieboczek/lifestolen/config/ClientConfig.kt
package nieboczek.lifestolen.config

import nieboczek.lifestolen.config.serializer.base.BooleanSerializer
import nieboczek.lifestolen.config.serializer.base.ClassSerializer
import nieboczek.lifestolen.config.serializer.base.ListSerializer
import nieboczek.lifestolen.config.serializer.base.StringSerializer

/** Access instance at [nieboczek.lifestolen.Lifestolen.cfg] */
class ClientConfig {
    // Populate with default values
    var renderClientBrandText = false
    var friends = mutableListOf<String>()
    var commandPrefix = "."
    var transformInput = true

    companion object {
        const val ID: String = "Client"
        val serializer = ClassSerializer { ClientConfig() }
            .field("RenderClientBrandText", BooleanSerializer(), { it.renderClientBrandText }, { c, v -> c.renderClientBrandText = v })
            .field("Friends", ListSerializer(StringSerializer()), { it.friends }, { c, v -> c.friends = v })
            .field("CommandPrefix", StringSerializer(), { it.commandPrefix }, { c, v -> c.commandPrefix = v })
            .field("TransformInput", BooleanSerializer(), { it.transformInput }, { c, v -> c.transformInput = v })
    }
}
